File: tests/Tests.FSharp/RangeSet.Tests.fs
module Zeta.Tests.RangeSetTests

open System.IO
open System.Reflection
open System.Text.Json
open global.Xunit
open Zeta.Core

// RangeSet — the F# oracle (#2 of TS/F#/C#/Rust). The TS reference
// (src/Core.TypeScript/range-set/) authors the shared golden vectors; this proves the F# impl
// replays them: render(parse input) == canonical (the cross-language byte lock) + contains agrees,
// and the rejection vectors decline the SPECIFIC feedback variant. "The compilers don't lie."

let private repoRoot () : string =
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location))

    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent

    if isNull dir then
        failwith "Could not locate repo root (Zeta.sln) from test assembly location."

    dir.FullName

let private golden () : JsonDocument =
    let path = Path.Join(repoRoot (), "src", "Core.TypeScript", "range-set", "golden-vectors.json")
    JsonDocument.Parse(File.ReadAllText path)

let private feedbackName (f: RangeSet.RangeSetFeedback) : string =
    match f with
    | RangeSet.NotInteger _ -> "NotInteger"
    | RangeSet.InvertedRange _ -> "InvertedRange"
    | RangeSet.Malformed _ -> "Malformed"

let private parseOk (input: string) : RangeSet.RangeSet =
    match RangeSet.parse input with
    | Ok rs -> rs
    | Error f -> failwithf "expected Ok for %A, got Error %A" input f

[<Fact>]
let ``F# RangeSet replays the shared golden cases (render(parse) == canonical + contains agrees)`` () =
    use doc = golden ()
    let cases = doc.RootElement.GetProperty("cases").EnumerateArray() |> Seq.toList
    Assert.True(cases.Length > 0, "no golden cases")

    for c in cases do
        let name = c.GetProperty("name").GetString()
        let input = c.GetProperty("input").GetString()
        let canonical = c.GetProperty("canonical").GetString()
        let rs = parseOk input
        Assert.Equal(canonical, RangeSet.render rs) // ($"{name}")
        // canonical is a fixed point of parse->render
        Assert.Equal(canonical, RangeSet.render (parseOk canonical))

        for probe in c.GetProperty("contains").EnumerateArray() do
            let arr = probe.EnumerateArray() |> Seq.toArray
            let n = arr.[0].GetInt64()
            let expected = arr.[1].GetBoolean()
            Assert.True((RangeSet.contains rs n) = expected, sprintf "%s: contains %d expected %b" name n expected)

[<Fact>]
let ``F# RangeSet rejection vectors decline the specific feedback variant`` () =
    use doc = golden ()

    for r in doc.RootElement.GetProperty("rejections").EnumerateArray() do
        let name = r.GetProperty("name").GetString()
        let input = r.GetProperty("input").GetString()
        let expected = r.GetProperty("feedback").GetString()

        match RangeSet.parse input with
        | Error f -> Assert.Equal(expected, feedbackName f)
        | Ok rs -> failwithf "%s: expected Error %s, got Ok %A" name expected rs

[<Fact>]
let ``F# RangeSet structural laws — union/add/size coalesce + count`` () =
    Assert.Equal("1-6", RangeSet.render (RangeSet.union (parseOk "1-3") (parseOk "4-6")))
    Assert.Equal("1-6,10-14", RangeSet.render (RangeSet.union (parseOk "1-5,10-12") (parseOk "6,13-14")))
    Assert.Equal("1-7", RangeSet.render (RangeSet.add (parseOk "1-3,5-7") 4L))
    Assert.Equal("1-3,10", RangeSet.render (RangeSet.add (parseOk "1-3") 10L))
    Assert.Equal(0L, RangeSet.size (parseOk ""))
    Assert.Equal(14L, RangeSet.size (parseOk "1-5,8,10-17"))

File: tests/Tests.CSharp/RangeSet/RangeSetTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp.RangeSet;

namespace Zeta.Tests.CSharp.RangeSet;

/// <summary>
/// RangeSet — the C# oracle (#3 of TS/F#/C#/Rust). The TS reference
/// (<c>src/Core.TypeScript/range-set/</c>) authors the shared golden vectors; this replays them:
/// <c>Render(Parse(input))</c> == canonical (the cross-language byte lock) + <c>Contains</c>
/// agrees, and the rejection vectors decline the SPECIFIC feedback variant. "The compilers don't lie."
/// </summary>
public class RangeSetTests
{
    private static string GoldenPath()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(RangeSetTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }

        var root = dir?.FullName
            ?? throw new InvalidOperationException("Could not locate repo root (Zeta.sln) from test assembly location.");
        return Path.Join(root, "src", "Core.TypeScript", "range-set", "golden-vectors.json");
    }

    private static IReadOnlyList<Interval> ParseOk(string input) =>
        Assert.IsType<Result<IReadOnlyList<Interval>, RangeSetFeedback>.Ok>(RangeSets.Parse(input)).Value;

    private static string FeedbackName(RangeSetFeedback f) => f switch
    {
        RangeSetFeedback.NotInteger => "NotInteger",
        RangeSetFeedback.InvertedRange => "InvertedRange",
        RangeSetFeedback.Malformed => "Malformed",
        _ => "?",
    };

    [Fact]
    public void ReplaysEverySharedGoldenCaseRenderEqualsCanonicalAndContainsAgrees()
    {
        using var doc = JsonDocument.Parse(File.ReadAllText(GoldenPath()));
        var cases = doc.RootElement.GetProperty("cases").EnumerateArray().ToList();
        Assert.NotEmpty(cases);

        foreach (var c in cases)
        {
            var input = c.GetProperty("input").GetString()!;
            var canonical = c.GetProperty("canonical").GetString()!;
            var rs = ParseOk(input);
            Assert.Equal(canonical, RangeSets.Render(rs));
            // canonical is a fixed point of parse->render
            Assert.Equal(canonical, RangeSets.Render(ParseOk(canonical)));

            foreach (var probe in c.GetProperty("contains").EnumerateArray())
            {
                var arr = probe.EnumerateArray().ToList();
                Assert.Equal(arr[1].GetBoolean(), RangeSets.Contains(rs, arr[0].GetInt64()));
            }
        }
    }

    [Fact]
    public void RejectionVectorsDeclineTheSpecificFeedbackVariant()
    {
        using var doc = JsonDocument.Parse(File.ReadAllText(GoldenPath()));
        foreach (var r in doc.RootElement.GetProperty("rejections").EnumerateArray())
        {
            var input = r.GetProperty("input").GetString()!;
            var expected = r.GetProperty("feedback").GetString()!;
            var err = Assert.IsType<Result<IReadOnlyList<Interval>, RangeSetFeedback>.Err>(RangeSets.Parse(input));
            Assert.Equal(expected, FeedbackName(err.Error));
        }
    }

    [Fact]
    public void StructuralLawsUnionAddSizeCoalesceAndCount()
    {
        Assert.Equal("1-6", RangeSets.Render(RangeSets.Union(ParseOk("1-3"), ParseOk("4-6"))));
        Assert.Equal("1-6,10-14", RangeSets.Render(RangeSets.Union(ParseOk("1-5,10-12"), ParseOk("6,13-14"))));
        Assert.Equal("1-7", RangeSets.Render(RangeSets.Add(ParseOk("1-3,5-7"), 4)));
        Assert.Equal("1-3,10", RangeSets.Render(RangeSets.Add(ParseOk("1-3"), 10)));
        Assert.Equal(0L, RangeSets.Size(ParseOk("")));
        Assert.Equal(14L, RangeSets.Size(ParseOk("1-5,8,10-17")));
    }

    [Fact]
    public void ParseNullDeclinesMalformedRatherThanThrowing()
    {
        // result-over-throw / no-exception-crosses-boundary: null must decline a feedback
        // variant, never NRE on s.Trim() (C#-specific parity; the other oracles are non-null by type).
        var err = Assert.IsType<Result<IReadOnlyList<Interval>, RangeSetFeedback>.Err>(RangeSets.Parse(null!));
        Assert.IsType<RangeSetFeedback.Malformed>(err.Error);
    }
}
